pytorch_toolkit/formula_recognition/im2latex/data/utils.py

```diff
@@ -0,0 +1,510 @@
+        return [cv.dilate(image, self.kernel, iterations=self.iterations) for image in img]
+
+
+class BatchTransformBin():
+    def __init__(self, threshold):
+
+        self.transform = cv.threshold
+        self.thresh_type = cv.THRESH_BINARY
+        self.threshold = threshold
+        self.max_val = 255
+
+    def __call__(self, img):
+        if not isinstance(img, list):
+            img = [img]
+        return [cv.threshold(im, self.threshold, self.max_val, self.thresh_type)[1] for im in img]
+
+
+class BatchTransfromAdaptiveBin():
+    def __init__(self, threshold, block_size, method=cv.ADAPTIVE_THRESH_MEAN_C, mean_c=10):
+        self.method = method
+        self.block_size = block_size
+        self.C = mean_c
+        self.thresh_type = cv.THRESH_BINARY
+        self.threshold = threshold
+        self.max_val = 255
+
+    def __call__(self, img):
+        if not isinstance(img, list):
+            img = [img]
+        img = [cv.cvtColor(im, cv.COLOR_BGR2GRAY) for im in img]
+        return [cv.adaptiveThreshold(
+            im, self.max_val, self.method, self.thresh_type,
+            self.block_size, self.C) for im in img]
+
+
+class BatchTransformRescale():
+    def __init__(self, scale_min, scale_max):
+        self.scale_min = scale_min
+        self.scale_max = scale_max
+
+    def __call__(self, imgs):
+        fx = np.random.uniform(self.scale_min, self.scale_max)
+        fy = fx  # np.random.uniform(fx, self.scale_max)
+        if not isinstance(imgs, list):
+            imgs = [imgs]
+        imgs = [cv.resize(img, dsize=None, fx=fx, fy=fy) for img in imgs]
+        return imgs
+
+
+class BatchTransformRotate():
+    def __init__(self, angle):
+        self.angle = angle
+
+    def __call__(self, imgs):
+        bound = self.angle
+        angle = np.random.uniform(low=0 - bound, high=bound)
+        if not isinstance(imgs, list):
+            imgs = [imgs]
+
+        img = imgs[0]
+        h, w = img.shape[0:2]
+        center = int(w / 2), int(h / 2)
+
+        M = cv.getRotationMatrix2D(center, angle=angle, scale=1)
+        src = np.array([[[0, 0]], [[w, 0]], [[0, h]], [[w, h]]])
+
+        new_pts = cv.transform(src, M)
+        top_left, top_right, bottom_left, bottom_right = new_pts.astype(int)
+
+        bb = cv.boundingRect(new_pts)
+        empty_img_shape = (bb[3], bb[2])
+        empty_img = 255 * np.ones(empty_img_shape)
+        shift_matrix = np.array([[0, 0, 0-bb[0]], [0, 0, 0 - bb[1]]])
+        M = M + shift_matrix
+        rotated = [cv.warpAffine(
+            img, M, empty_img_shape[::-1], borderMode=cv.BORDER_CONSTANT, borderValue=COLOR_WHITE) for img in imgs]
+
+        return rotated
+
+
+def get_timestamp():
+    return "{:%Y-%m-%d_%H-%M-%S}".format(datetime.datetime.now())
+
+
+def get_num_lines_in_file(path):
+    total = 0
+    with open(path) as f:
+        for line in f:
+            total += 1
+    return total
+
+
+def cal_loss(logits, targets, should_cut_by_min=False):
```

Comment:
what is `cal`?

Reply:
calculate

Reply:
if I asks something it means that in most cases I know answer, but it should be fixed/re-done/re-renamed/imporved

Reply:
BTW, Vladimir, why this function (calculation of loss) is in the file `data/utils.py`    
It is used only in `tools/train.py` -- why not to move it to that file?

Reply:
Renamed & moved function